src/path_calculator.rs

```diff
@@ -57,30 +57,68 @@ impl PathCalculator {
         graph: &FastGraph,
         start: NodeId,
         end: NodeId,
+    ) -> Option<ShortestPath> {
+        self.calc_path_multiple_sources_and_targets(graph, vec![(start, 0)], vec![(end, 0)])
+    }
+
+    pub fn calc_path_multiple_sources_and_targets(
+        &mut self,
+        graph: &FastGraph,
+        starts: Vec<(NodeId, Weight)>,
+        ends: Vec<(NodeId, Weight)>,
     ) -> Option<ShortestPath> {
         assert_eq!(
             graph.get_num_nodes(),
             self.num_nodes,
             "given graph has invalid node count"
         );
-        assert!(start < self.num_nodes, "invalid start node");
-        assert!(end < self.num_nodes, "invalid end node");
+        assert!(starts.len() > 0, "there has to be at least one start");
+        assert!(ends.len() > 0, "there has to be at least one end");
+        for (start_node, _) in &starts {
+            assert!(*start_node < self.num_nodes, "invalid start node");
+        }
+        for (end_node, _) in &ends {
+            assert!(*end_node < self.num_nodes, "invalid end node");
+        }
         self.heap_fwd.clear();
         self.heap_bwd.clear();
         self.valid_flags_fwd.invalidate_all();
         self.valid_flags_bwd.invalidate_all();
-        if start == end {
-            return Some(ShortestPath::singular(start));
-        }
-
-        self.update_node_fwd(start, 0, INVALID_NODE, INVALID_EDGE);
-        self.update_node_bwd(end, 0, INVALID_NODE, INVALID_EDGE);
-        self.heap_fwd.push(HeapItem::new(0, start));
-        self.heap_bwd.push(HeapItem::new(0, end));
 
         let mut best_weight = WEIGHT_MAX;
         let mut meeting_node = INVALID_NODE;
 
+        for (start_node, start_weight) in &starts {
+            for (end_node, end_weight) in &ends {
+                if *start_node == *end_node
+                    && *start_weight < WEIGHT_MAX
+                    && *end_weight < WEIGHT_MAX
+                {
+                    if *start_weight + *end_weight < best_weight {
+                        best_weight = *start_weight + *end_weight;
+                        meeting_node = *end_node;
+                    }
+                }
+            }
+        }
+
+        for (node, weight) in starts {
```

Comment:
Just a minor thing: Here you used `for (...) in starts`, but above you used `for (...) in &starts`. Should we prefer one over the other here or do we actually need both of them?

Reply:
`&starts` just borrows the `Vec<(NodeId, Weight)>`, which is why you have to dereference `*end_weight` and similar. When you do `for (...) in starts {`, that consumes the vector, meaning it can't be used again after this. If you tried to iterate over `starts` twice, the compiler wouldn't let you -- the first iteration consumes the vector. It's slightly more performant to consume the vector than dereference everything inside. So the way it's written here is fine.

Also, [clippy](https://github.com/rust-lang/rust-clippy#usage) has made lots of progress since a year ago. It uses static analysis to find places where you could consume a vector instead of borrowing, and it helps with conventional Rust style. All of the warnings have good explanations for why to switch. Giving it a quick run on this crate, there are a bunch of easy things we could fix up. I'd be happy to send a quick PR after we merge this one.

Reply:
Ok thanks for the explanation. In my naive thinking borrowing was actually cheaper because it 'avoids a copy'. I'll definitely take a look into clippy and run it on this project :+1:, let me try first so I can learn something from it :)

Reply:
We could play with https://godbolt.org/ to know for sure, but my understanding is that more compiler optimizations might be available if a vector of primitive values is being consumed instead of borrowed. When borrowing, that's when you might need to make a copy.